```
Error running mint automation
+ mkdir -p 15516-5a6c149/gopath/src/github.com/minio
+ git -C 15516-5a6c149/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 15516-5a6c149/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 15516-5a6c149/gopath/src/github.com/minio/minio fetch --quiet minio pull/15516/head:pr15516
+ git -C 15516-5a6c149/gopath/src/github.com/minio/minio checkout --quiet pr15516
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/15516-5a6c149/gopath
+ make -C 15516-5a6c149/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/minio/pkg v1.3.0
go: downloading github.com/minio/minio-go/v7 v7.0.34
go: downloading github.com/minio/cli v1.22.0
go: downloading golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/klauspost/compress v1.15.9
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/Shopify/sarama v1.35.0
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/golang-jwt/jwt/v4 v4.4.1
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading golang.org/x/crypto v0.0.0-20220722155217-630584e8d5aa
go: downloading github.com/minio/kes v0.20.0
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/felixge/fgprof v0.9.2
go: downloading github.com/klauspost/reedsolomon v1.10.0
go: downloading github.com/go-openapi/loads v0.21.1
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/minio/console v0.19.4
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/miekg/dns v1.1.48
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/prometheus/client_golang v1.12.2
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/common v0.34.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/shirou/gopsutil/v3 v3.22.6
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.4
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5
go: downloading google.golang.org/api v0.78.0
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.1
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.10.4
go: downloading github.com/nats-io/nats.go v1.16.0
go: downloading github.com/nats-io/stan.go v0.10.2
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/beevik/ntp v0.3.0
go: downloading go.uber.org/zap v1.21.0
go: downloading golang.org/x/time v0.0.0-20220224211638-0e9765cccd65
go: downloading github.com/go-ldap/ldap/v3 v3.4.3
go: downloading github.com/coredns/coredns v1.9.2
go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
go: downloading github.com/mattn/go-colorable v0.1.12
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/fraugster/parquet-go v0.10.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.5.1
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/klauspost/cpuid/v2 v2.1.0
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/golang/snappy v0.0.4
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.4
go: downloading google.golang.org/grpc v1.46.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading google.golang.org/genproto v0.0.0-20220429170224-98d788798c3e
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20211209120228-48547f28849e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/go-openapi/analysis v0.21.2
go: downloading github.com/go-openapi/spec v0.20.5
go: downloading github.com/go-openapi/swag v0.21.1
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/google/pprof v0.0.0-20211214055906-6f57359322fd
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.0
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading cloud.google.com/go v0.100.2
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/rs/xid v1.4.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading gopkg.in/ini.v1 v1.66.6
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/lestrrat-go/jwx v1.2.19
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/pierrec/lz4/v4 v4.1.15
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/go-openapi/strfmt v0.21.2
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/mitchellh/mapstructure v1.4.3
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading go.mongodb.org/mongo-driver v1.8.4
go: downloading github.com/oklog/ulid v1.3.1
go: downloading github.com/go-openapi/errors v0.20.2
go: downloading cloud.google.com/go/iam v0.2.0
go: downloading github.com/googleapis/gax-go/v2 v2.3.0
go: downloading go.opencensus.io v0.23.0
go: downloading golang.org/x/xerrors v0.0.0-20220411194840-2f41105eb62f
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-stack/stack v1.8.1
go: downloading github.com/go-openapi/runtime v0.23.3
go: downloading github.com/go-openapi/validate v0.21.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/minio/mc v0.0.0-20220805080128-351d021b924b
go: downloading github.com/unrolled/secure v1.10.0
go: downloading github.com/tidwall/gjson v1.14.0
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/docker/go-units v0.4.0
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/pkg/xattr v0.4.5
go: downloading github.com/charmbracelet/lipgloss v0.5.0
go: downloading github.com/charmbracelet/bubbles v0.10.3
go: downloading github.com/minio/colorjson v1.0.2
go: downloading github.com/charmbracelet/bubbletea v0.20.0
go: downloading github.com/rivo/tview v0.0.0-20220216162559-96063d6082f3
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/navidys/tvxwidgets v0.1.0
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/posener/complete v1.2.3
go: downloading github.com/prometheus/prom2json v1.3.1
go: downloading github.com/gdamore/tcell/v2 v2.4.1-0.20210905002822-f057f0a857a1
go: downloading github.com/tklauser/go-sysconf v0.3.10
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/muesli/termenv v0.11.1-0.20220212125758-44cd13922739
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/tklauser/numcpus v0.4.0
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading cloud.google.com/go/compute v1.6.1
../../../../pkg/mod/github.com/minio/console@v0.19.4/pkg/logger/config/config.go:20:2: missing go.sum entry for module providing package github.com/minio/madmin-go (imported by github.com/minio/minio/cmd); to add:
	go get github.com/minio/minio/cmd
cmd/server-rlimit.go:24:2: missing go.sum entry for module providing package github.com/minio/madmin-go/kernel (imported by github.com/minio/minio/cmd); to add:
	go get github.com/minio/minio/cmd
make: *** [build] Error 1

```

--- a/cmd/admin-bucket-handlers.go
+++ b/cmd/admin-bucket-handlers.go
@@ -249,7 +249,7 @@ func (a adminAPIHandlers) SetRemoteTargetHandler(w http.ResponseWriter, r *http.
 	}
 	if err = globalBucketTargetSys.SetTarget(ctx, bucket, &target, update); err != nil {
 		switch err.(type) {
-		case BucketRemoteConnectionErr:
+		case RemoteTargetConnectionErr:
 			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErrWithErr(ErrReplicationRemoteConnectionError, err), r.URL)
 		default:
 			writeErrorResponseJSON(ctx, w, toAPIError(ctx, err), r.URL)

--- a/cmd/api-errors.go
+++ b/cmd/api-errors.go
@@ -888,7 +888,7 @@ var errorCodes = errorCodeMap{
 	},
 	ErrReplicationRemoteConnectionError: {
 		Code:           "XMinioAdminReplicationRemoteConnectionError",
-		Description:    "Remote service connection error - please check remote service credentials and target bucket",
+		Description:    "Remote service connection error",
 		HTTPStatusCode: http.StatusNotFound,
 	},
 	ErrReplicationBandwidthLimitError: {
@@ -2074,7 +2074,7 @@ func toAPIErrorCode(ctx context.Context, err error) (apiErr APIErrorCode) {
 		apiErr = ErrRemoteDestinationNotFoundError
 	case BucketRemoteTargetNotFound:
 		apiErr = ErrRemoteTargetNotFoundError
-	case BucketRemoteConnectionErr:
+	case RemoteTargetConnectionErr:
 		apiErr = ErrReplicationRemoteConnectionError
 	case BucketRemoteAlreadyExists:
 		apiErr = ErrBucketRemoteAlreadyExists

--- a/cmd/bucket-replication-handlers.go
+++ b/cmd/bucket-replication-handlers.go
@@ -310,7 +310,7 @@ func (api objectAPIHandlers) ResetBucketReplicationStartHandler(w http.ResponseW
 	rinfo.Targets = append(rinfo.Targets, ResyncTarget{Arn: tgtArns[0], ResetID: target.ResetID})
 	if err = globalBucketTargetSys.SetTarget(ctx, bucket, &target, true); err != nil {
 		switch err.(type) {
-		case BucketRemoteConnectionErr:
+		case RemoteTargetConnectionErr:
 			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErrWithErr(ErrReplicationRemoteConnectionError, err), r.URL)
 		default:
 			writeErrorResponseJSON(ctx, w, toAPIError(ctx, err), r.URL)

--- a/cmd/bucket-replication.go
+++ b/cmd/bucket-replication.go
@@ -534,7 +534,7 @@ func replicateDeleteToTarget(ctx context.Context, dobj DeletedObjectReplicationI
 	if dobj.VersionID != "" && rinfo.VersionPurgeStatus == Complete {
 		return
 	}
-	if tgt.IsOffline() {
+	if globalBucketTargetSys.isOffline(tgt.EndpointURL()) {
 		logger.LogIf(ctx, fmt.Errorf("remote target is offline for bucket:%s arn:%s", dobj.Bucket, tgt.ARN))
 		sendEvent(eventArgs{
 			BucketName: dobj.Bucket,
@@ -1039,7 +1039,7 @@ func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, object
 		rinfo.ReplicationResynced = true
 		return
 	}
-	if tgt.IsOffline() {
+	if globalBucketTargetSys.isOffline(tgt.EndpointURL()) {
 		logger.LogIf(ctx, fmt.Errorf("remote target is offline for bucket:%s arn:%s", bucket, tgt.ARN))
 		sendEvent(eventArgs{
 			EventName:  event.ObjectReplicationNotTracked,
@@ -1659,7 +1659,7 @@ func proxyHeadToRepTarget(ctx context.Context, bucket, object string, rs *HTTPRa
 	}
 	for _, t := range proxyTargets.Targets {
 		tgt = globalBucketTargetSys.GetRemoteTargetClient(ctx, t.Arn)
-		if tgt == nil || tgt.IsOffline() {
+		if tgt == nil || globalBucketTargetSys.isOffline(tgt.EndpointURL()) {
 			continue
 		}
 		// if proxying explicitly disabled on remote target

--- a/cmd/bucket-targets.go
+++ b/cmd/bucket-targets.go
@@ -19,6 +19,8 @@ package cmd
 
 import (
 	"context"
+	"fmt"
+	"net/url"
 	"sync"
 	"time"
 
@@ -35,7 +37,9 @@ import (
 )
 
 const (
-	defaultHealthCheckDuration = 30 * time.Second
+	defaultHealthCheckDuration = 5 * time.Second
+	// default interval for reload of all remote target endpoints
+	defaultHealthCheckReloadDuration = 30 * time.Minute
 )
 
 // BucketTargetSys represents bucket targets subsystem
@@ -43,6 +47,113 @@ type BucketTargetSys struct {
 	sync.RWMutex
 	arnRemotesMap map[string]*TargetClient
 	targetsMap    map[string][]madmin.BucketTarget
+	hMutex        sync.RWMutex
+	hc            map[string]epHealth
+	hcClient      *madmin.AnonymousClient
+}
+
+// epHealth struct represents health of a replication target endpoint.
+type epHealth struct {
+	Endpoint string
+	Scheme   string
+	Online   bool
+}
+
+// isOffline returns current liveness result of remote target. Add endpoint to
+// healthcheck map if missing and default to online status
+func (sys *BucketTargetSys) isOffline(ep *url.URL) bool {
+	sys.hMutex.RLock()
+	defer sys.hMutex.RUnlock()
+	if h, ok := sys.hc[ep.Host]; ok {
+		return !h.Online
+	}
+	go sys.initHC(ep)
+	return false
+}
+
+func (sys *BucketTargetSys) initHC(ep *url.URL) {
+	sys.hMutex.Lock()
+	sys.hc[ep.Host] = epHealth{
+		Endpoint: ep.Host,
+		Scheme:   ep.Scheme,
+		Online:   true,
+	}
+	sys.hMutex.Unlock()
+}
+
+// newHCClient initializes an anonymous client for performing health check on the remote endpoints
+func newHCClient() *madmin.AnonymousClient {
+	clnt, e := madmin.NewAnonymousClientNoEndpoint()
+	if e != nil {
+		logger.LogOnceIf(GlobalContext, fmt.Errorf("WARNING: Unable to initialize health check client"), string(replicationSubsystem))
+		return nil
+	}
+	clnt.SetCustomTransport(globalRemoteTargetTransport)
+	return clnt
+}
+
+// heartBeat performs liveness check on remote endpoints.
+func (sys *BucketTargetSys) heartBeat(ctx context.Context) {
+	hcTimer := time.NewTimer(defaultHealthCheckDuration)
+	defer hcTimer.Stop()
+	for {
+		select {
+		case <-hcTimer.C:
+			sys.hMutex.RLock()
+			var eps []madmin.ServerProperties
+			for _, ep := range sys.hc {
+				eps = append(eps, madmin.ServerProperties{Endpoint: ep.Endpoint, Scheme: ep.Scheme})
+			}
+			sys.hMutex.RUnlock()
+
+			if len(eps) > 0 {
+				cctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+				defer cancel()
+				m := map[string]epHealth{}
+				for result := range sys.hcClient.Alive(cctx, madmin.AliveOpts{}, eps...) {
+					var online bool
+					if result.Error == nil {
+						online = result.Online
+					}
+					m[result.Endpoint.Host] = epHealth{
+						Endpoint: result.Endpoint.Host,
+						Scheme:   result.Endpoint.Scheme,
+						Online:   online,
+					}
+				}
+				sys.hMutex.Lock()
+				sys.hc = m
+				sys.hMutex.Unlock()
+			}
+			hcTimer.Reset(defaultHealthCheckDuration)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// periodically rebuild the healthcheck map from list of targets to clear
+// out stale endpoints
+func (sys *BucketTargetSys) reloadHealthCheckers(ctx context.Context) {
+	m := make(map[string]epHealth)
+	tgts := sys.ListTargets(ctx, "", "")
+	for _, t := range tgts {
+		if _, ok := m[t.Endpoint]; !ok {
+			scheme := "http"
+			if t.Secure {
+				scheme = "https"
+			}
+			m[t.Endpoint] = epHealth{
+				Online:   true,
+				Endpoint: t.Endpoint,
+				Scheme:   scheme,
+			}
+		}
+	}
+	sys.hMutex.Lock()
+	// swap out the map
+	sys.hc = m
+	sys.hMutex.Unlock()
 }
 
 // ListTargets lists bucket targets across tenant or for individual bucket, and returns
@@ -91,9 +202,6 @@ func (sys *BucketTargetSys) Delete(bucket string) {
 		return
 	}
 	for _, t := range tgts {
-		if tgt, ok := sys.arnRemotesMap[t.Arn]; ok && tgt.healthCancelFn != nil {
-			tgt.healthCancelFn()
-		}
 		delete(sys.arnRemotesMap, t.Arn)
 	}
 	delete(sys.targetsMap, bucket)
@@ -116,7 +224,7 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
 		if minio.ToErrorResponse(err).Code == "NoSuchBucket" {
 			return BucketRemoteTargetNotFound{Bucket: tgt.TargetBucket}
 		}
-		return BucketRemoteConnectionErr{Bucket: tgt.TargetBucket, Err: err}
+		return RemoteTargetConnectionErr{Bucket: tgt.TargetBucket, Err: err}
 	}
 	if tgt.Type == madmin.ReplicationService {
 		if !globalBucketVersioningSys.Enabled(bucket) {
@@ -124,7 +232,7 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
 		}
 		vcfg, err := clnt.GetBucketVersioning(ctx, tgt.TargetBucket)
 		if err != nil {
-			return BucketRemoteConnectionErr{Bucket: tgt.TargetBucket, Err: err}
+			return RemoteTargetConnectionErr{Bucket: tgt.TargetBucket, Err: err}
 		}
 		if vcfg.Status != string(versioning.Enabled) {
 			return BucketRemoteTargetNotVersioned{Bucket: tgt.TargetBucket}
@@ -152,10 +260,7 @@ func (sys *BucketTargetSys) SetTarget(ctx context.Context, bucket string, tgt *m
 	if !found && !update {
 		newtgts = append(newtgts, *tgt)
 	}
-	// cancel health check for previous target client to avoid leak.
-	if prevClnt, ok := sys.arnRemotesMap[tgt.Arn]; ok && prevClnt.healthCancelFn != nil {
-		prevClnt.healthCancelFn()
-	}
+
 	sys.targetsMap[bucket] = newtgts
 	sys.arnRemotesMap[tgt.Arn] = clnt
 	sys.updateBandwidthLimit(bucket, tgt.BandwidthLimit)
@@ -227,9 +332,6 @@ func (sys *BucketTargetSys) RemoveTarget(ctx context.Context, bucket, arnStr str
 		return BucketRemoteTargetNotFound{Bucket: bucket}
 	}
 	sys.targetsMap[bucket] = targets
-	if tgt, ok := sys.arnRemotesMap[arnStr]; ok && tgt.healthCancelFn != nil {
-		tgt.healthCancelFn()
-	}
 	delete(sys.arnRemotesMap, arnStr)
 	sys.updateBandwidthLimit(bucket, 0)
 	return nil
@@ -258,11 +360,29 @@ func (sys *BucketTargetSys) GetRemoteBucketTargetByArn(ctx context.Context, buck
 }
 
 // NewBucketTargetSys - creates new replication system.
-func NewBucketTargetSys() *BucketTargetSys {
-	return &BucketTargetSys{
+func NewBucketTargetSys(ctx context.Context) *BucketTargetSys {
+	sys := &BucketTargetSys{
 		arnRemotesMap: make(map[string]*TargetClient),
 		targetsMap:    make(map[string][]madmin.BucketTarget),
+		hc:            make(map[string]epHealth),
+		hcClient:      newHCClient(),
 	}
+	// reload healthcheck endpoints map periodically to remove stale endpoints from the map.
+	go func() {
+		rTimer := time.NewTimer(defaultHealthCheckReloadDuration)
+		defer rTimer.Stop()
+		for {
+			select {
+			case <-rTimer.C:
+				sys.reloadHealthCheckers(ctx)
+				rTimer.Reset(defaultHealthCheckReloadDuration)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	go sys.heartBeat(ctx)
+	return sys
 }
 
 // UpdateAllTargets updates target to reflect metadata updates
@@ -276,9 +396,6 @@ func (sys *BucketTargetSys) UpdateAllTargets(bucket string, tgts *madmin.BucketT
 	// Remove existingtarget and arn association
 	if tgts, ok := sys.targetsMap[bucket]; ok {
 		for _, t := range tgts {
-			if tgt, ok := sys.arnRemotesMap[t.Arn]; ok && tgt.healthCancelFn != nil {
-				tgt.healthCancelFn()
-			}
 			delete(sys.arnRemotesMap, t.Arn)
 		}
 		delete(sys.targetsMap, bucket)
@@ -345,10 +462,6 @@ func (sys *BucketTargetSys) getRemoteTargetClient(tcfg *madmin.BucketTarget) (*T
 	if tcfg.HealthCheckDuration >= 1 { // require minimum health check duration of 1 sec.
 		hcDuration = tcfg.HealthCheckDuration
 	}
-	cancelFn, err := api.HealthCheck(hcDuration)
-	if err != nil {
-		return nil, err
-	}
 	tc := &TargetClient{
 		Client:              api,
 		healthCheckDuration: hcDuration,
@@ -356,9 +469,10 @@ func (sys *BucketTargetSys) getRemoteTargetClient(tcfg *madmin.BucketTarget) (*T
 		Bucket:              tcfg.TargetBucket,
 		StorageClass:        tcfg.StorageClass,
 		disableProxy:        tcfg.DisableProxy,
-		healthCancelFn:      cancelFn,
 		ARN:                 tcfg.Arn,
 		ResetID:             tcfg.ResetID,
+		Endpoint:            tcfg.Endpoint,
+		Secure:              tcfg.Secure,
 	}
 	return tc, nil
 }
@@ -432,7 +546,8 @@ type TargetClient struct {
 	replicateSync       bool
 	StorageClass        string // storage class on remote
 	disableProxy        bool
-	healthCancelFn      context.CancelFunc // cancellation function for client healthcheck
-	ARN                 string             // ARN to uniquely identify remote target
+	ARN                 string // ARN to uniquely identify remote target
 	ResetID             string
+	Endpoint            string
+	Secure              bool
 }

--- a/cmd/object-api-errors.go
+++ b/cmd/object-api-errors.go
@@ -422,11 +422,18 @@ func (e BucketRemoteTargetNotFound) Error() string {
 	return "Remote target not found: " + e.Bucket
 }
 
-// BucketRemoteConnectionErr remote target connection failure.
-type BucketRemoteConnectionErr GenericError
+// RemoteTargetConnectionErr remote target connection failure.
+type RemoteTargetConnectionErr struct {
+	Err      error
+	Bucket   string
+	Endpoint string
+}
 
-func (e BucketRemoteConnectionErr) Error() string {
-	return fmt.Sprintf("Remote service endpoint or target bucket not available: %s \n\t%s", e.Bucket, e.Err.Error())
+func (e RemoteTargetConnectionErr) Error() string {
+	if e.Bucket != "" {
+		return fmt.Sprintf("Remote service endpoint offline or target bucket/remote service credentials invalid: %s \n\t%s", e.Bucket, e.Err.Error())
+	}
+	return fmt.Sprintf("Remote service endpoint %s not available\n\t%s", e.Endpoint, e.Err.Error())
 }
 
 // BucketRemoteAlreadyExists remote already exists for this target type.

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -290,7 +290,7 @@ func initAllSubsystems() {
 	}
 
 	// Create new bucket replication subsytem
-	globalBucketTargetSys = NewBucketTargetSys()
+	globalBucketTargetSys = NewBucketTargetSys(GlobalContext)
 
 	// Create new ILM tier configuration subsystem
 	globalTierConfigMgr = NewTierConfigMgr()

--- a/cmd/site-replication.go
+++ b/cmd/site-replication.go
@@ -2217,6 +2217,9 @@ func getAdminClient(endpoint, accessKey, secretKey string) (*madmin.AdminClient,
 	if err != nil {
 		return nil, err
 	}
+	if globalBucketTargetSys.isOffline(epURL) {
+		return nil, RemoteTargetConnectionErr{Endpoint: epURL.String(), Err: fmt.Errorf("remote target is offline for endpoint %s", epURL.String())}
+	}
 	client, err := madmin.New(epURL.Host, accessKey, secretKey, epURL.Scheme == "https")
 	if err != nil {
 		return nil, err
@@ -2230,6 +2233,10 @@ func getS3Client(pc madmin.PeerSite) (*minioClient.Client, error) {
 	if err != nil {
 		return nil, err
 	}
+	if globalBucketTargetSys.isOffline(ep) {
+		return nil, RemoteTargetConnectionErr{Endpoint: ep.String(), Err: fmt.Errorf("remote target is offline for endpoint %s", ep.String())}
+	}
+
 	return minioClient.New(ep.Host, &minioClient.Options{
 		Creds:     credentials.NewStaticV4(pc.AccessKey, pc.SecretKey, ""),
 		Secure:    ep.Scheme == "https",

--- a/go.mod
+++ b/go.mod
@@ -48,7 +48,7 @@ require (
 	github.com/minio/dperf v0.4.2
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes v0.20.0
-	github.com/minio/madmin-go v1.4.17
+	github.com/minio/madmin-go v1.4.20
 	github.com/minio/minio-go/v7 v7.0.34
 	github.com/minio/pkg v1.3.0
 	github.com/minio/selfupdate v0.5.0

